vignettes/miaViz.Rmd

```diff
@@ -326,6 +326,127 @@ names(metadata(dmn_se))
+```
+
+Principal Coordinate Analysis using Bray-Curtis dissimilarity on the
+`hitchip1006` dataset:
+
+```{r}
+library(dplyr)
+data(hitchip1006, package = "miaTime")
+tse <- hitchip1006
+tse <- transformCounts(tse, method = "relabundance")
+## Ordination with PCoA with Bray-Curtis dissimilarity
+tse <- runMDS(tse, FUN = vegan::vegdist, method = "bray",
+                  name = "PCoA_BC", exprs_values = "relabundance",
+              na.rm = TRUE)
+# Calculating Explained variance
+e <- attr(reducedDim(tse, "PCoA_BC"), "eig")
+rel_eig <- e/sum(e[e>0])
+# plot
+p1 <- plotReducedDim(tse, dimred = "PCoA_BC") +
+    labs(x = paste("PCoA 1 (", round(100 * rel_eig[[1]],1), "%", ")", sep = ""),
+         y = paste("PCoA 2 (", round(100 * rel_eig[[2]],1), "%", ")", sep = ""),
+         title="PCoA - Bray-Curtis")+
+    theme(plot.title = element_text(hjust = 0.5))
+p1
+```
+
+Retrieving information about all available trajectories:
+
+```{r}
+colData(tse)$time <- as.factor(colData(tse)$time)
+# trajectories info
+traj_info <- table(colData(tse)$subject,colData(tse)$time)
+# has at least two points
+traj_info <- traj_info[rowSums(traj_info)!=1,]
+table(rowSums(traj_info)) %>% as.data.frame() %>%
+    `colnames<-`(c("Trajectories Lengths", "Counts"))
+```
+
+Lets look at all present trajectories in the data:
+
+```{r}
+# keeping only data that has one-to-one link from one point to another
+df <- traj_info %>% as.data.frame() %>%
+    `colnames<-`(c("subject", "time", "value")) %>% 
+    filter(value ==1) %>% arrange(subject, time) 
+# retrieving corresponding ordination coordinate values
+df <- cbind(df, 
+      reducedDim(
+          tse[,(colData(tse)$subject %in% df$subject &
+                    colData(tse)$time %in% df$time)], "PCoA_BC") %>%
+          `colnames<-`(c("PC1","PC2")) %>% as_tibble(rownames = "sample_id") %>% 
+          mutate(subject=colData(tse)[sample_id,"subject"]) %>%
+          arrange(subject)) %>% select(-last_col())
+# retrieving the corresponding start and end of each trajectory 
+df_start_end <- rbind(
+    df %>% group_by(subject) %>% slice(which.min(time)) %>% ungroup() %>%
+        mutate(tp=rep("start",nrow(.))),
+    df %>% group_by(subject) %>% slice(which.max(time)) %>% ungroup() %>%
+        mutate(tp=rep("end",nrow(.)))
+)
+# plot
+p1 + 
+    geom_path(aes(x=PC1, y=PC2, color=subject),
+              data = df )+ 
+    geom_point(aes(x=PC1,y=PC2, shape=tp),
+               data = df_start_end ,
+               size=2)+
+    scale_shape_manual(name="",values = c("start"=8, "end"=5))+
+    labs(title = "All present trajectories")
+```
+
+Filtering by trajectory length and displaying top 10%:
+
+```{r}
+# calculating trajectory lengths and merge
+df <- merge(df,
+               df %>% group_by(subject) %>% group_modify(~{
+                   mat <- as.matrix(dist(.x[,c("PC1", "PC2")]));
+                   data.frame(change=sum(mat[row(mat) == col(mat) + 1]))
+               }))
```

Comment:
it calculates the euclidean distance (based on the ordination coordinates) between each pair of successive time points; if a trajectory has more than two time points it calculates the sum of the euclidean distance between each successive time points.
I will add comments.

Reply:
Shall I keep the entire examples also only for two timpoints? or rather show all available trajectories present in the data as per current implementation?

Reply:
update: the method suggested at the earlier example `selected.subjects <- names(which(table(tse$subject)==2))` would also work for all available trajectories --> `selected.subjects <- names(which(table(tse$subject)>=2))` shall keep both examples of two time points and all trajectories available?

Reply:
Note that the miaTime package includes function [getStepwiseDivergence.html](https://microbiome.github.io/miaTime/reference/getStepwiseDivergence.html) for calculating distances between consecutive time points. I think it could apply here as well with small modifications (using PC coordinates instead of the original data). If yes, it would be very useful to take advantage of that, and avoid duplicating the effort with different code.

Reply:
For now case with two time points is perhaps sufficient. Although it might be informative to include later another plot that has for instance one subject with a longer trajectory highlighted.

Reply:
I checked the getStepwiseDivergence but won't pick data only from assays, and reducedDim data cannot be assigned to assays. As a workaround, I created another TreeSE with assay as the reducedDim of tse and colData from same. getStepwiseDivergence calculate only with vegan::vegdist(method="bray"), not working when passing to the ... arguments method="euclidean". Doing another workaround for it still throws the following error `Error in as.character.factor(X[[i]], ...) : malformed factor`. This is the only mod I could figure out the moment to be able use to use getStepwiseDivergence, totalling in much more code than just the way how it's done above.

Reply:
ok, I will do two time point and another plot with longest trajectories.

Reply:
Interesting! 

Then the question is shall I/we update `getStepwiseDivergence` function so that one could call assays, altExps, and reducedDims, and use other distance methods.. 

For now perhaps the fastest option is not to wait for that but just implement the simplest possible example. However, this might be worth looking at immediately afterwards.

Reply:
https://github.com/microbiome/miaTime/issues/64

Reply:
Here is a reproducible example that's working atm, however it take a lot of time to compute the distances/change, perhaps filtering out subjects having only one time point before passing it to  `getStepwiseDivergence` could speed it up:

```
library(miaTime)
library(mia)
library(scater)
library(dplyr)

data("hitchip1006")
tse <- hitchip1006
tse <- transformCounts(tse, method = "relabundance")
## Ordination with PCoA with Bray-Curtis dissimilarity
tse <- runMDS(tse, FUN = vegan::vegdist, method = "bray",
                  name = "PCoA_BC", exprs_values = "relabundance",
              na.rm = TRUE)
tse_tmp <- TreeSummarizedExperiment(
    assay=SimpleList(PCoA=reducedDim(tse, "PCoA_BC") %>% as.data.frame() %>% 
                         `colnames<-`(c("PC_1", "PC_2")) %>% t),
    colData=colData(tse)
)

tse_tmp <- getStepwiseDivergence(tse_tmp, group = "subject",
                                 time_field = "time",
                                 assay_name = "PCoA",
                                 FUN = function(x) {vegan::vegdist(x, method = "euclidean")})
colData(tse_tmp) %>% as_tibble() %>% filter(!is.na(time_difference)) %>%
        group_by(subject) %>% summarise(change=sum(time_divergence))
```

Reply:
Another thing that helps to speed up is to filter down to the most prevalent taxa.

Some more comments in the code will always help.

Is `getStepwiseDivergence` really only working with "euclidean" distance? Can you open an issue in miaTime reporting this? It does make sense for this specific visualization task in 2D ordination space, however.

Ok here we create reducedDim, then we use that to create a new TreeSE; and on that we calculate stepwise divergences. Summarise seems to calculate the overall trajectory length, I am not sure what is the purpose of that.

Reply:
`getStepwiseDivergence` will work only with "bray" when I tested it, does not parse/use the argument `method="euclidean"` when given to `...` (presumably I thought it would be further given to `vegan::vegdist`).
The purpose of the summarise is to calculate the sum of euclidean distances between each successive points to get the whole trajectory length, to be used as a way of filtering trajectories by lengths in the subsequent examples.

Reply:
Could you open an issue to miaTime and describe this issue? I think this is important to solve somehow.